Update databricks deployments to support AI gateway & additional update endpoints

djliden opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/djliden/mlflow/pull/13513?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13513/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13513
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->

#13344 

### What changes are proposed in this pull request?

1. Updates the `create_endpoint` method to accept the whole API request in the config argument, rather than just the `config` component of the payload, so now users can include e.g. the AI gateway configuration, tags, etc.
2. Adds update methods corresponding to the api endpoints for updating config, tags, gateway, rate limits
3. Adds new tests for the new & updated methods

### How is this PR tested?

- [X] Existing unit/integration tests
- [X] New unit/integration tests
- [X] Manual tests

```python
# Create
from mlflow.deployments import get_deploy_client

client = get_deploy_client("databricks")
endpoint = client.create_endpoint(
    config={
        "name": "test",
        "config": {
            "served_entities": [
                {
                    "external_model": {
                        "name": "gpt-4",
                        "provider": "openai",
                        "task": "llm/v1/chat",
                        "openai_config": {
                            "openai_api_key": "{{secrets/scope/key}}",
                        },
                    },
                }
            ],
            "route_optimized": True,
        },
    },
)

# update config

name = "test"
config = {
    "served_entities": [
        {
            "name": "gpt-4o-mini",
            "external_model": {
                "name": "gpt-4o-mini",
                "provider": "openai",
                "task": "llm/v1/chat",
                "openai_config": {"openai_api_key_plaintext": os.getenv("OPENAI_API_KEY")},
            },
        }
    ]
}

updated_endpoint = client.update_endpoint_config(endpoint=name, config=config)

updated_endpoint
# {'name': 'test',
#  'creator': 'daniel.liden@databricks.com',
#  'creation_timestamp': 1729527763000,
#  'last_updated_timestamp': 1729531568000,
#  'state': {'ready': 'READY', 'config_update': 'NOT_UPDATING'},
#  'config': {'served_entities': [{'name': 'gpt-4o-mini',
#     'type': 'EXTERNAL_MODEL',
#     'external_model': {'provider': 'openai',
#      'name': 'gpt-4o-mini',
#      'task': 'llm/v1/chat',
#      'openai_config': {}},
#     'state': {'deployment': 'DEPLOYMENT_READY',
#      'deployment_state_message': ''},
#     'creator': 'daniel.liden@databricks.com',
#     'creation_timestamp': 1729531568000}],
#   'traffic_config': {'routes': [{'served_model_name': 'gpt-4o-mini',
#      'traffic_percentage': 100}]},
#   'config_version': 1},
#  'id': '44b258fb39804564b37603d8d14b853e',
#  'permission_level': 'CAN_MANAGE',
#  'tags': [{'key': 'project', 'value': 'test'}],
#  'route_optimized': False,
#  'task': 'llm/v1/chat',
#  'endpoint_type': 'EXTERNAL_MODEL',
#  'creator_display_name': 'Daniel Liden',
#  'creator_kind': 'User'}

# update Tags
name = "test"
tags = {"add_tags": [{"key": "project", "value": "test"}]}
updated_tags = client.update_endpoint_tags(endpoint=name, config=tags)

updated_tags

# {'tags': [{'key': 'project', 'value': 'test'}]}

# Rate Limits
name = "databricks-dbrx-instruct"
rate_limits = {"rate_limits": [{"calls": 10, "key": "endpoint", "renewal_period": "minute"}]}
updated_rate_limits = client.update_endpoint_rate_limits(endpoint=name, config=rate_limits)
updated_rate_limits

# {'rate_limits': [{'calls': 10, 'key': 'endpoint', 'renewal_period': 'minute'}]}

# Update ai gateway

name = "test"
gateway_config = {
    "usage_tracking_config": {"enabled": True},
    "inference_table_config": {
        "enabled": True,
        "catalog_name": CATALOG_NAME,
        "schema_name": SCHEMA_NAME,
    },
}
updated_gateway = client.update_endpoint_ai_gateway(endpoint=name, config=gateway_config)
updated_gateway

name = "test"

# gateway_config = {
#     "usage_tracking_config": {"enabled": True},
#     "inference_table_config": {
#         "enabled": True,
#         "catalog_name": CATALOG_NAME,
#         "schema_name": SCHEMA_NAME,
#     },
# }

updated_gateway = client.update_endpoint_ai_gateway(endpoint=name, config=gateway_config)

updated_gateway

```
### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [X] Examples
  - [X] API references
  - [ ] Instructions

(Just the automatically-generated documentation)

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Adds support for configuring Mosaic AI Gateway in Databricks Model Serving to MLflow Deployments. Adds support for passing entire API request payload to `create_endpoint`. Adds new `update_endpoint_xxx` methods corresponding to the Databricks model serving update endpoints.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

djliden committed:
updates create_endpoint method

create_endpoint now expects the full API request payload as the config.

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
updates docstring with example of config-as-full-payload approach

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
returns error if name argument used w full payload approach

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
Adds update endpoint methods

adds specific methods for updating config, ai-gateway, rate-limits,
and tags, consistent with the various update endpoints in the
databricks model serving API

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
updates tests for create/update methods

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
ruff auto-format

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
updates warning for old create_endpoint args

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
ruff autoformat

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden commented:
cc @BenWilson2 to check for any databricks-specific tests

BenWilson2 commented on mlflow/deployments/databricks/__init__.py:
```diff
@@ -351,42 +352,67 @@ def create_endpoint(self, name, config=None, route_optimized=False):
-        payload = {"name": name, "config": config, "route_optimized": route_optimized, **extras}
+        if config and "config" in config:
+            # config contains the full API request payload
+            if route_optimized or name:
```
Is this safe? 

What if I define `name` at the top level arg and don't put it in the `config` kwargs? That would be the old way of defining the endpoint and we shouldn't break that experience. 

We can update this to:
1. Check if `name` is defined in `config`. 
a. If it is, check that name matches the arg `name`. 
    I. If they match, issue a deprecation warning about the usage of the top-level `name` argument. 
    II. If they do not match, raise an MLflowException explaining that both cannot be defined and that `name` should be defined within config, messaging that the `name` named arg is deprecated. 
b. If it is not, then ensure that `name` is set (it is a required arg that is now synthetically made optional) and raise an MLflowException if it is not set, indicating to users to supply it within `config`. 

For the `route_optimized` argument, similar changes will be needed, but due to the default value that used to exist, the logic will need to be slightly different.

djliden replied:
I think the earlier version was safe and enforced a stricter separation between the "old" and "new" approaches—it would, for example, error if a user put name as a param and in the config, even if they were the same.

I updated according to your suggestions—I think the approach of checking if they are the same and issuing a warning, and only erroring if there is a name mismatch, is more forgiving.

BenWilson2 commented on mlflow/deployments/databricks/__init__.py:
```diff
@@ -351,42 +352,67 @@ def create_endpoint(self, name, config=None, route_optimized=False):
+            payload = config
+        else:
+            # for backwards compatibility
+            warnings.warn(
```
See above comment for how we should restructure the deprecation warning here (the messages will likely need to be appended to a list and separate messages for the individual fields will be required to be emitted in the warning message).

BenWilson2 commented on mlflow/deployments/databricks/__init__.py:
```diff
@@ -335,10 +336,10 @@ def create_endpoint(self, name, config=None, route_optimized=False):
         Args:
             name: The name of the serving endpoint to create.
-            config: A dictionary containing the configuration of the serving endpoint to create.
+            config: A dictionary containing either the full API request payload
```
We'll need a deprecation indicator for the APIs that we don't want people using anymore so that this notice appears in the API docs

BenWilson2 commented on tests/deployments/databricks/test_databricks.py:
```diff
@@ -46,6 +46,38 @@ def test_create_endpoint():
         assert resp == {"name": "test"}
 
 
+def test_create_endpoint_config_only():
```
Can we add tests for the error handling logic that needs to be implemented for backwards compatibility / validation of the deprecated state? 
We need to verify that setting `name` in both `config` and as a named arg throws. 
We need to verify that setting `route_optimized` in both will throw.
We need to ensure that the correct warning is emitted when both are provided but the values match. 
We need to ensure that a deprecation warning is emitted when named args are used and they are not set in the config (separate warning from the previous row)

harupy commented:
@BenWilson2 Do you think we still need `DatabricksDeploymentClient`? I'm wondering if we can just use databricks SDK.

djliden commented:
@harupy @BenWilson2 given that, should I close this PR given, or still worth making these changes in the meantime?

FWIW the SDK approach to setting gateway and other options ends up being a lot more verbose and fairly challenging to configure given how strongly typed it is—e.g a fairly basic external model ends up looking like this

```
create_endpoint_config = EndpointCoreConfigInput(
    name="dr-gateway-demo",
    served_entities=[
        ServedEntityInput(
            name="<name>",
            external_model=ExternalModel(
                name="<model-name>",
                provider=ExternalModelProvider.OPENAI,
                task="llm/v1/chat",
                openai_config=OpenAiConfig(
                    openai_api_key=f"{{{{secrets/{SECRETS_SCOPE}/{SECRETS_KEY}}}}}",
                ),
            ),
        )
    ],
)
```
with a lot of nested classes, especially when you get into gateway, rate limits, etc. The MLflow deployments approach was a bit easier to use in that respect. Fully understand if that's something to take up with the databricks sdk and not a justification for having an MLflow implementation, though.

BenWilson2 commented:
@djliden yeah that is pretty terrible and likely won't really get used for no other reason than "you can't do this without looking in the docs for examples" and IIRC there isn't any guidance or examples of how to do something like this with the auto-generated SDK docs. 
Let's update this PR and we can consider whether to make a light wrapper around the Workspace API later.

djliden committed:
adds deprecation flags

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
updates checks for name and route_optimized

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
adds deprecation warning to update_config

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
updates tests to check warnings and errors

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
fixes deprecation warning formatting

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden committed:
applies ruff formatting

Signed-off-by: djliden <7102904+djliden@users.noreply.github.com>

djliden commented:
Updated deprecation warnings and added new tests of the various pathways through the `create_endpoint` method. I tried to follow the format of other deprecation flags in the codebase, but please do let me know if there is a preferred approach I missed.

BenWilson2 reviewed (approved):
LGTM!
